### ISSUE TITLE
Found with fuzzer: You turn into a demon orc!  You shudder for a moment.

### ISSUE THREAD

**Issue by elunna** [elunna/hackem#22]
```
=================================================================
==10883==ERROR: AddressSanitizer: heap-use-after-free on address 0x60c0006b3533 at pc 0x5555557bafa7 bp 0x7fffffffe230 sp 0x7fffffffe220
READ of size 1 at 0x60c0006b3533 thread T0
    #0 0x5555557bafa6 in doapply /home/lunatunez/hackem/src/apply.c:5039
    #1 0x55555586306e in rhack /home/lunatunez/hackem/src/cmd.c:5517
    #2 0x555555763bff in moveloop /home/lunatunez/hackem/src/allmain.c:798
    #3 0x5555563bb425 in main ../sys/unix/unixmain.c:353
    #4 0x7ffff7367082 in __libc_start_main ../csu/libc-start.c:308
    #5 0x55555575ab0d in _start (/home/lunatunez/games/hackemdir/hackem+0x206b0d)

0x60c0006b3533 is located 51 bytes inside of 120-byte region [0x60c0006b3500,0x60c0006b3578)
freed by thread T0 here:
    #0 0x7ffff769b40f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x555555d52eee in dealloc_obj /home/lunatunez/hackem/src/mkobj.c:2483
    #2 0x5555560b2b05 in obfree /home/lunatunez/hackem/src/shk.c:1159
    #3 0x555555af538b in useupall /home/lunatunez/hackem/src/invent.c:1120
    #4 0x555555af5499 in useup /home/lunatunez/hackem/src/invent.c:1135
    #5 0x555555794009 in use_mask /home/lunatunez/hackem/src/apply.c:2519
    #6 0x55555594f9f7 in Blindf_on /home/lunatunez/hackem/src/do_wear.c:1801
    #7 0x5555557b92f9 in doapply /home/lunatunez/hackem/src/apply.c:4806
    #8 0x55555586306e in rhack /home/lunatunez/hackem/src/cmd.c:5517
    #9 0x555555763bff in moveloop /home/lunatunez/hackem/src/allmain.c:798
    #10 0x5555563bb425 in main ../sys/unix/unixmain.c:353
    #11 0x7ffff7367082 in __libc_start_main ../csu/libc-start.c:308

previously allocated by thread T0 here:
    #0 0x7ffff769b808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x555555766f7e in alloc /home/lunatunez/hackem/src/alloc.c:46
    #2 0x555555d4670a in mksobj /home/lunatunez/hackem/src/mkobj.c:806
    #3 0x555555d42ca0 in mkobj /home/lunatunez/hackem/src/mkobj.c:287
    #4 0x555555dc56ee in xkilled /home/lunatunez/hackem/src/mon.c:4238
    #5 0x555555dc17f1 in killed /home/lunatunez/hackem/src/mon.c:4116
    #6 0x5555562731d3 in hmon_hitmon /home/lunatunez/hackem/src/uhitm.c:1990
    #7 0x555556255c33 in hmon /home/lunatunez/hackem/src/uhitm.c:915
    #8 0x555556252ab3 in known_hitum /home/lunatunez/hackem/src/uhitm.c:634
    #9 0x55555625446a in hitum /home/lunatunez/hackem/src/uhitm.c:791
    #10 0x55555624f542 in attack /home/lunatunez/hackem/src/uhitm.c:579
    #11 0x555555ac7639 in domove_core /home/lunatunez/hackem/src/hack.c:1813
    #12 0x555555ab1001 in domove /home/lunatunez/hackem/src/hack.c:1516
    #13 0x555555862b5f in rhack /home/lunatunez/hackem/src/cmd.c:5479
    #14 0x555555763bff in moveloop /home/lunatunez/hackem/src/allmain.c:798
    #15 0x5555563bb425 in main ../sys/unix/unixmain.c:353
    #16 0x7ffff7367082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-use-after-free /home/lunatunez/hackem/src/apply.c:5039 in doapply
Shadow bytes around the buggy address:
  0x0c18800ce650: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c18800ce660: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c18800ce670: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c18800ce680: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c18800ce690: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
=>0x0c18800ce6a0: fd fd fd fd fd fd[fd]fd fd fd fd fd fd fd fd fa
  0x0c18800ce6b0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c18800ce6c0: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c18800ce6d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c18800ce6e0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c18800ce6f0: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==10883==ABORTING
```

**Comment by elunna** [comment]
<img width="960" alt="issue 22" src="https://user-images.githubusercontent.com/5000096/198116495-0e9f603b-0d43-402d-a92d-068049f0d4ec.png">

**Comment by elunna** [comment]
Problem with splicehack masks

If the mask is cursed, we suffer a system shock from polymorph.
I can reproduce this by applying a cursed mask of a demon orc.

**Comment by elunna** [comment]
Fixed in commit 42841e516.When masks broke, obj was being checked later in doapply to see if it could speak. Since it was deallocated, this caused a heap-use-after-free. Now when masks are applied, they return 1 immediately and skip the check at the end.
